Tools/ci-run.sh

```diff
@@ -3,125 +3,164 @@
+    if [[ $BACKEND == *"cpp"* ]]; then
       echo "WARNING: Currently not installing pythran due to compatibility issues"
       # python -m pip install pythran==0.9.5 || exit 1
-    fi
-
-    if [ "$BACKEND" != "cpp" -a -n "${PYTHON_VERSION##pypy*}" -a -n "${PYTHON_VERSION##2*}" -a -n "${PYTHON_VERSION##*3.4}" ]; then
+    elif [[ $PYTHON_VERSION == "pypy"* && $PYTHON_VERSION == "2"* && $PYTHON_VERSION == *"3.4" ]]; then
       python -m pip install mypy || exit 1
     fi
   fi
 fi
 
 # Run tests
+echo "==== Running tests ===="
 ccache -s 2>/dev/null || true
 export PATH="/usr/lib/ccache:$PATH"
 
-if [ "$NO_CYTHON_COMPILE" != "1" -a -n "${PYTHON_VERSION##pypy*}" ]; then
-  CFLAGS="-O2 -ggdb -Wall -Wextra $(python -c 'import sys; print("-fno-strict-aliasing" if sys.version_info[0] == 2 else "")')" \
-  python setup.py build_ext -i \
-          $(if [ "$COVERAGE" == "1" ]; then echo " --cython-coverage"; fi) \
-          $(if [ "$CYTHON_COMPILE_ALL" == "1" ]; then echo " --cython-compile-all"; fi) \
-          $(python -c 'import sys; print("-j5" if sys.version_info >= (3,5) else "")') \
-      || exit 1
-  if [ -z "$COVERAGE" -a -z "$STACKLESS" -a -z "$LIMITED_API" -a -z "$CYTHON_COMPILE_ALL" -a -z "$EXTRA_CFLAGS" -a -n "${BACKEND//*cpp*}" ]; then
+if [[ "$OSTYPE" == "msys" ]]; then  # for MSVC cl
+  # /wd will disable warning
+  # 4711 warns that function `x` selected for automatic inline expansion
+  # 4127 warns that conditional expression is constant, should be fixed here https://github.com/cython/cython/pull/4317
+  # (off by default) 5045 warns that Compiler will insert Spectre mitigation for memory load if /Qspectre switch specified
+  # (off by default) 4820 warns about the code in Python\3.9.6\x64\include ...
+  WARNARGS="/W4 /wd4711 /wd4127 /wd5045 /wd4820"
+  DEBUG_INFO="/Z7"
+  NO_OPTIMIZATION="-Od"
+else
+  WARNARGS="-Wall -Wextra"
+  DEBUG_INFO="-ggdb"
+  NO_OPTIMIZATION="-O0"
+fi
+
+if [[ $NO_CYTHON_COMPILE != "1" && $PYTHON_VERSION == "pypy"* ]]; then
+  if [[ $COVERAGE == "1" ]]; then
+    COVERAGE_ARGS="--cython-coverage"
+  fi
+  if [[ $CYTHON_COMPILE_ALL == "1" ]]; then
+    CYTHON_COMPILE_ALL_FLAGS=" --cython-compile-all"
+  fi
+  if [[ $PYTHON_SYS_VERSION > "3.5" || $PYTHON_SYS_VERSION == "3.5"* ]]; then
+    SETUP_ARGS="-j5"
+  fi
+  if [[ $PYTHON_SYS_VERSION == "2"* ]]; then
+    ALIASING="$-fno-strict-aliasing"
+  fi
+
+  CFLAGS="-O2 $DEBUG_INFO $WARNARGS $ALIASING" \
+    python setup.py build_ext -i \
+    $COVERAGE_ARGS $CYTHON_COMPILE_ALL_FLAGS $SETUP_ARGS || exit 1
+
+  if [[ $COVERAGE != "1" && $STACKLESS != "true" && -z $LIMITED_API && $CYTHON_COMPILE_ALL != "1" && -z $EXTRA_CFLAGS && $BACKEND != *"cpp"* ]]; then
     python setup.py bdist_wheel || exit 1
   fi
 fi
 
-if [ "$TEST_CODE_STYLE" == "1" ]; then
-    make -C docs html || exit 1
-elif [ -n "${PYTHON_VERSION##pypy*}" ]; then
+if [[ $TEST_CODE_STYLE == "1" ]]; then
+  make -C docs html || exit 1
+elif [[ $PYTHON_VERSION == "pypy"* ]]; then
```

Comment:
@scoder is [this](https://github.com/mattip/cython/commit/1a783ca8e6566f2157ff2b3e643093dfbb9b498f#commitcomment-54108949) relevant?

Reply:
Yes. Note that you reversed the condition above. The debugger tests should *not* run on on PyPy (nor Windows).

Reply:
> Note that you reversed the condition above. The debugger tests should _not_ run on on PyPy (nor Windows).

Yeah, my bad

Reply:
In your original [comment](https://github.com/mattip/cython/commit/1a783ca8e6566f2157ff2b3e643093dfbb9b498f#commitcomment-54108949) you used `"${OS_NAME##windows*}" == ""` which will be true when the OS is Windows, but then you said that we shouldn't run the debugger on Winsows, so I think you made a mistake and forgot to write `-n`, so I'll use `$OSTYPE != "msys"`

Reply:
[resolved](https://github.com/cython/cython/pull/4324/commits/ff16175ff3de0fbd86af1b06402ae78a06882e72)